test/form-component-spec.js

```diff
@@ -3,1609 +3,1710 @@ import { assert } from 'chai';
+        inputElement.value = 'valid';
+        TestUtils.Simulate.change(inputElement);
 
-      assert.ok(store.getState().testForm.$form.valid);
+        assert.isTrue(store.getState().testForm.$form.valid);
+      });
     });
-  });
 
-  describe('form revalidation after manual validation', () => {
-    const store = testCreateStore({
-      login: modelReducer('login', { foo: 'bar' }),
-      loginForm: formReducer('login', { foo: 'bar' }),
-    });
-    const handleSubmit = (values) => {
-      if (values.foo !== 'changed') {
-        store.dispatch(actions.setValidity('login', {
-          correctDetails: false,
-        }));
+    describe('invalidating async validity on form change with form validators', () => {
+      const initialState = getInitialState({ foo: 'invalid' });
+      const store = createTestStore(testCreateStore({
+        test: modelReducer('test', initialState),
+        testForm: formReducer('test', initialState),
+      }));
+
+      function handleSubmit() {
+        store.dispatch(actions.batch('test', [
+          actions.setSubmitFailed('test'),
+          actions.setErrors('test', 'Form is invalid', { errors: true }),
+        ]));
       }
-    };
-    const form = testRender(
-      <Form model="login" onSubmit={handleSubmit}>
-        <Field model="login.foo">
-          <input />
-        </Field>
-      </Form>, store);
 
-    const input = TestUtils.findRenderedDOMComponentWithTag(form, 'input');
-    const formNode = TestUtils.findRenderedDOMComponentWithTag(form, 'form');
+      const form = TestUtils.renderIntoDocument(
+        <Provider store={store}>
+          <Form
+            model="test"
+            validators={{
+              foo: (val) => val && val.length,
+            }}
+            onSubmit={handleSubmit}
+          >
+            <Field model="test.foo">
+              <input type="text" />
+            </Field>
+          </Form>
+        </Provider>
+      );
 
-    it('should revalidate after being set invalid', () => {
-      TestUtils.Simulate.submit(formNode);
+      const formElement = TestUtils.findRenderedDOMComponentWithTag(form, 'form');
+      const inputElement = TestUtils.findRenderedDOMComponentWithTag(form, 'input');
 
-      assert.isFalse(store.getState().loginForm.$form.valid);
+      it('should set errors from rejected submit handler on valid submit', () => {
+        assert.isTrue(store.getState().testForm.$form.valid);
 
-      input.value = 'changed';
+        TestUtils.Simulate.submit(formElement);
 
-      TestUtils.Simulate.change(input);
+        assert.containSubset(
+          store.getState().testForm.$form,
+          { errors: 'Form is invalid' });
+      });
 
-      TestUtils.Simulate.submit(formNode);
+      it('should set validity on form changes after submit failed', () => {
+        inputElement.value = 'valid';
+        TestUtils.Simulate.change(inputElement);
 
-      assert.isTrue(store.getState().loginForm.$form.valid);
+        assert.isT
```

Comment:
Below is the test I added for the bug fix in handleSubmit() changing the check on formValue.valid to formValue.$form.valid.  It may seem a little weird at first glance because we already have tests covering that the form calls the submit handler when it's valid but the goal was to exercise this block of code in handleSubmit():

```
const formValid = formValue
        ? formValue.$form.valid
        : true;

      if (!validators && onSubmit && formValid) {
        onSubmit(modelValue);

        return modelValue;
      }
```

which was being totally skipped in all cases previously.  So the tests below were the best way I could think to cover that but please feel free to let me know if you have any other ideas!